### PR TITLE
Add additional tests for the `core` module

### PR DESCRIPTION
**Pull request by M-Lampert**
This PR adds additional tests and fixes some minor formatting and typing issues as well as some bugs that have been encountered while writing the tests.

### CODE REVIEW

File: tests/core/test_dag_data.py [hunk 1]
```diff
@@ -169,50 +101,54 @@ def test_edge_index_kth_order_dag():
...
+        IntTensor([[[0, 1, 2], [0, 1, 2]], [[1, 2, 3], [1, 2, 4]]]).to(config["torch"]["device"]),
     )
 
+    assert equal(w3, tensor([1.0, 1.0]).to(config["torch"]["device"]))  # a-b-c -> b-c-d | a-b-c -> b-c-e
 
-def test_lift_order_dag():
-    e1 = tensor([[[0], [1], [1], [3]], [[1], [2], [3], [4]]])
-    x = DAGData.lift_order_dag(e1)
-    assert equal(x, IntTensor([[[0, 1], [0, 1], [1, 3]], [[1, 2], [1, 3], [3, 4]]]))
+    with pytest.raises(ValueError):
```
Comment (M-Lampert) [inline tests/core/test_dag_data.py:145]: This is why `test_edge_index_k_weighted` currently fails. I.e. if you choose an order that is larger than the length of the longest `DAG` the method returns a RunTimeError. Should this stay this way or should we instead return a more meaningful error message or just an empty tensor?

File: tests/core/test_graph.py [hunk 2]
```diff
@@ -1,54 +1,367 @@
...
+    idxs = [str(i) for i in range(100)]
+    mapping = IndexMap(idxs)
+    g = Graph(data, mapping)
+    assert isinstance(g, Graph)
+    assert isinstance(g.data, Data)
+    assert isinstance(g.mapping, IndexMap)
+    assert isinstance(g.edge_to_index, dict)
+
+
+def test_init_with_list_mapping():
```
Comment (M-Lampert) [inline tests/core/test_graph.py:48]: Should we support initializing a `Graph` like this?

File: tests/core/test_graph.py [hunk 3]
```diff
@@ -1,54 +1,367 @@
...
+        ("a", "c"),
+    ]
+    g = Graph.from_edge_list(edge_list, is_undirected=True)
+    assert isinstance(g, Graph)
+    assert isinstance(g.data, Data)
+    assert isinstance(g.mapping, IndexMap)
+    assert isinstance(g.edge_to_index, dict)
+
+
+def test_from_edge_list_fails():
```
Comment (M-Lampert) [inline tests/core/test_graph.py:97]: Is this a behaviour that we want or do we trust the users to only input correct things?

File: tests/core/test_graph.py [hunk 4]
```diff
@@ -1,54 +1,367 @@
...
 
+    s = [v for v in simple_graph.predecessors("a")]
+    assert len(s) == 0
+
 
 def test_is_edge(simple_graph):
     assert simple_graph.is_edge("a", "b")
+    assert not simple_graph.is_edge("b", "a")
+    assert simple_graph.is_edge("a", "c")
+    assert not simple_graph.is_edge("c", "a")
```
Comment (M-Lampert) [inline tests/core/test_graph.py:192]: TODO: Needs fixing

File: tests/core/test_graph.py [hunk 5]
```diff
@@ -1,54 +1,367 @@
...
+    assert laplacian.data[4] == 1
+    assert laplacian.data[5] == 0
+
+
+def test_add_node_ohe(simple_graph):
+    simple_graph.add_node_ohe("node_ohe")
+    assert simple_graph.data["node_ohe"].shape == (3, 3)
+    assert torch.equal(simple_graph.data["node_ohe"], torch.eye(3))
+
+    simple_graph.add_node_ohe("node_ohe_dim_small", 2)
```
Comment (M-Lampert) [inline tests/core/test_graph.py:257]: What is the benefit of specifying `dim` here? Is the below behaviour something that we want?

File: tests/core/test_graph.py [hunk 6]
```diff
@@ -1,54 +1,367 @@
...
+    assert simple_graph["edge_weight", "a", "c"].item() == 2
+
+
+def test_get_graph_attr(simple_graph):
+    simple_graph.data["graph_feature"] = torch.tensor([42])
+    assert simple_graph["graph_feature"].item() == 42
+
+
+def test_set_node_attr(simple_graph):
+    simple_graph["node_class"] = torch.tensor([[1], [2], [3]])
```
Comment (M-Lampert) [inline tests/core/test_graph.py:306]: Should assignment like this be possible?

File: tests/core/test_temporal_graph.py [hunk 7]
```diff
@@ -35,3 +70,14 @@ def test_get_window(long_temporal_graph):
     # N defined by largest node index!
     assert t_2.N == 8
     assert t_2.M == 4
+
+
+def test_get_snapshot(long_temporal_graph):
```
Comment (M-Lampert) [inline tests/core/test_temporal_graph.py:75]: This is a bug from merging. Still TODO!

File: tests/core/test_dag_data.py (hunk 1 above)
Comment (IngoScholtes) [reply tests/core/test_dag_data.py:145]: We should return an error. Need to see how and where to catch this ...

File: tests/core/test_graph.py (hunk 2 above)
Comment (IngoScholtes) [reply tests/core/test_graph.py:48]: Yes, this should be the default constructor for all core classes, i.e. initializing from a `pyG.data.Data` object, with additional static class methods to create instances from other data structures (which are mapped to `pyG.data.Data`,

File: tests/core/test_graph.py (hunk 3 above)
Comment (IngoScholtes) [reply tests/core/test_graph.py:97]: This is the intended behavior, i.e. we require the user to give correct input, raising an exception if the input is wrong.

File: tests/core/test_graph.py (hunk 4 above)
Comment (IngoScholtes) [reply tests/core/test_graph.py:192]: This issue is due to the CSR format returned by pyG.EdgeIndex for matrix with rows that only contain zeros. 

My understanding of the CSR format is that a zero row would be represented by a repeated entry in the `row_ptr`, which allows to recover the zero row. The `get_csr` function of `EdgeIndex` seems to omits this. This also seems to be an issue when calling the `EdgeIndex.to_dense()` function, which returns a dense, possibly non-square matrix where all zero rows are missing, i.e. for the example above it returns 

```
tensor([[0., 1., 1.],
        [0., 0., 1.]])
```

instead of 

```
tensor([[0., 1., 1.],
        [0., 0., 1.],
        [0., 0., 0]])
```

Need to check whether this behavior of the pyG CSR format is intended.

File: tests/core/test_graph.py (hunk 5 above)
Comment (IngoScholtes) [reply tests/core/test_graph.py:257]: The idea was that we can specify the dimension of the one-hot encoding. However, we can probably remove this function, as the user can now directly access attributes in the underlying `Data` frame, which allows to add one hot encodings as well.

File: tests/core/test_graph.py (hunk 6 above)
Comment (IngoScholtes) [reply tests/core/test_graph.py:306]: Yes, that should be possible.
